File: Data Structures & Algorithms/implement-prefix-tree/submission-1.java
class PrefixTree {
    Node head;
    public PrefixTree() {
        head = new Node();
    }

    public void insert(String word) {
        Node cur = head;
        for (char c: word.toCharArray()) {
            Node temp = cur.children.get(c);
            if (temp == null) {
                temp = new Node();
                cur.children.put(c, temp);
            }
            cur = temp;
        }
        cur.end = true;
    }

    public boolean search(String word) {
        Node cur = head;
        for (char c: word.toCharArray()) {
            Node temp = cur.children.get(c);
            if (temp == null) {
                return false;
            }
            cur = temp;
        }
        return cur.end;
    }

    public boolean startsWith(String prefix) {
        Node cur = head;
        for (char c: prefix.toCharArray()) {
            Node temp = cur.children.get(c);
            if (temp == null) {
                return false;
            }
            cur = temp;
        }
        return true;
    }

     
    class Node {
        HashMap<Character, Node> children;
        public boolean end;

        public Node() {
            this.children = new HashMap<Character, Node>();
            end = false;
        }
    }
}

File: Data Structures & Algorithms/design-word-search-data-structure/submission-1.py
class WordDictionary:

    def __init__(self):
        self.head = dict()

    def addWord(self, word: str) -> None:
        cur = self.head
        for i in list(word):
            if i not in cur:
                cur[i] = {0: False}
            cur = cur[i]
        cur[0] = True

    def search(self, word: str) -> bool:

        def path(index, cur) -> bool:
            if index == len(word):
                return cur[0]
            if word[index] == '.':
                for nextc in cur:
                    if nextc == 0:
                        continue
                    if path(index + 1, cur[nextc]):
                        return True
            else:
                if word[index] in cur:
                    return path(index + 1, cur[word[index]])
            return False

        return path(0, self.head)